go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,544 @@
+		avoidResync,
+	)
+	netAttachDefInformer := nadFactory.K8sCniCncfIo().V1().NetworkAttachmentDefinitions()
+	rateLimter := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(qps), qps*5)})
+
+	nadController := &netAttachDefinitionController{
+		ncm:                ncm,
+		nadFactory:         nadFactory,
+		netAttachDefLister: netAttachDefInformer.Lister(),
+		netAttachDefSynced: netAttachDefInformer.Informer().HasSynced,
+		queue:              workqueue.NewNamedRateLimitingQueue(rateLimter, "net-attach-def"),
+		loopPeriod:         time.Second,
+
+		NetAttachDefManager: &netAttachDefManager{
+			perNadNetConfInfo:     syncmap.NewSyncMap[*nadNetConfInfo](),
+			perNetworkNadNameInfo: syncmap.NewSyncMap[*nadNameInfo](),
+		},
+	}
+	netAttachDefInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    nadController.onNetworkAttachDefinitionAdd,
+			UpdateFunc: nadController.onNetworkAttachDefinitionUpdate,
+			DeleteFunc: nadController.onNetworkAttachDefinitionDelete,
+		})
+	return nadController
+}
+
+func (nadController *netAttachDefinitionController) Run(stopChan <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer nadController.queue.ShutDown()
+
+	nadController.nadFactory.Start(stopChan)
+	klog.Infof("Starting controller %s", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopChan, nadController.netAttachDefSynced) {
+		return fmt.Errorf("error syncing cache")
+	}
+
+	err := nadController.repairNads()
+	if err != nil {
+		klog.Errorf("Failed to sync all existing nad entries: %v", err)
+	}
+
+	klog.Info("Starting workers for controller %s", controllerName)
+	wg := &sync.WaitGroup{}
+	for i := 0; i < numberOfWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.Until(nadController.worker, nadController.loopPeriod, stopChan)
+		}()
+	}
+	wg.Wait()
+
+	// wait until we're told to stop
+	<-stopChan
+
+	klog.Infof("Shutting down controller %s", controllerName)
+	nadController.queue.ShutDown()
+	return nil
+}
+
+func (nadController *netAttachDefinitionController) repairNads() (err error) {
+	startTime := time.Now()
+	klog.V(4).Infof("Starting repairing loop for %s", controllerName)
+	defer func() {
+		klog.V(4).Infof("Finished repairing loop for %s: %v err: %v", controllerName,
+			time.Since(startTime), err)
+	}()
+
+	existingNads, err := nadController.netAttachDefLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to get list of all net-attach-def")
+	}
+
+	// need to walk through all the nads and update nadNames of Controller. Controllers will be created as
+	// the result, but they can only be started afterwards when all nads are added to the network controller.
+	for _, nad := range existingNads {
+		_ = nadController.AddNetAttachDef(nadController.ncm, nad, false)
+	}
```

Comment:
but this is needed to initialize all the nads of all the networks without starting the controller, otherwise, when the network controller is created and started when first nad is added, not all nads are ready in the controller which would result in error.

Reply:
OK, I see. 

I understand that the errors would be temporary but that things would eventually succeed. We are not talking about permanent errors?

Can this be explained here better in the comment? Something like "Start the controller after all known NADs are added so as to avoid to the extent possible the errors and retries that would result if the controller attempted to process pods attached with NADs we wouldn't otherwise know about yet".